flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -535,13 +536,16 @@ public void checkpointStopped(long checkpointId) {
 	/**
 	 * Returns a list of buffers, checking the first n non-priority buffers, and skipping all events.
 	 */
-	private List<Buffer> getInflightBuffersUnsafe(long checkpointId) {
+	private List<Buffer> getInflightBuffersUnsafe(long checkpointId) throws CheckpointException {
 		assert Thread.holdsLock(receivedBuffers);
 
+		if (checkpointId < lastBarrierId) {
+			throw new CheckpointException(
```

Comment:
I was in doubt whether this exception type is appropriate here or whether an `Optional` should be returned.
But at the upper level, it ultimately should be used, and the reason is only known here (so it would be harder to analyze the exception/empty buffers later).

Reply:
I think this is exactly the case for which `CheckpointException` was designed.